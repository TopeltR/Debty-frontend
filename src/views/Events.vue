<template>
    <div>
        <navbar/>
        <background>
            <spinner :loaded="loaded">
                <div v-if="events && events.length > 0">
                    <h1 class='pt-4'>My events
                        <font-awesome-icon icon='plus' class=' ml-3 mt-2 limegreen' v-on:click='createNewEvent'/>
                    </h1>
                    <div class='search-wrapper mr-4 mt-4 mb-4'>
                        <b-row>
                            <b-col md="6">
                                <input class='form-control' type='text' v-model='search' placeholder='Search'/>
                            </b-col>
                        </b-row>
                    </div>
                    <div v-for="event in filteredList" :key="event.id" @click="goToEvent(event.id)"
                         class="card my-3 shadow">
                        <div class="card-body">
                            <b-row>
                                <b-col md="6">
                                    <h4 class="ml-2">{{event.title}}</h4>
                                </b-col>
                                <b-col md="6" class="d-none d-md-block text-right">
                                    <b>Created: </b> {{getDateString(event.createdAt)}}
                                    <span v-if="event.closedAt">
                                    <br>
                                    <b>Closed: </b> {{getDateString(event.closedAt)}}
                                    </span>
                                </b-col>
                            </b-row>
                            <b-row class="mx-2">
                                <small><b>Owner: </b> {{getFullName(event.owner)}}</small>
                            </b-row>
                            <b-row v-if="event.description.length > 0" class="mx-2 mt-3">
                                <p>{{event.description}}</p>
                            </b-row>
                        </div>
                    </div>
                </div>
                <div v-else class="text-center p-5">
                    <h1>You don't have any events yet, grab your friends and create one!</h1>
                    <b-btn class="m-5 btn-limegreen px-3 py-2" v-on:click="goToEventCreation">Create event!</b-btn>
                </div>
            </spinner>
        </background>
    </div>
</template>

<script>
    import Navbar from '@/components/Navbar.vue';
    import Background from '@/components/Background.vue';
    import Spinner from '@/components/Spinner.vue';
    import router from '@/router';
    import userStore from '@/stores/UserStore';
    import eventStore from '@/stores/EventStore';

    export default {
        name: 'Events',
        components: {Navbar, Background, Spinner},

        mounted() {
            this.getEvents();
        },
        data() {
            return {
                events: undefined,
                search: '',
            };
        },
        methods: {
            createNewEvent() {
                router.push('/events/create');
            },
            goToEvent(id) {
                router.push('/events/' + id);
            },
            sortByTime() {
                this.events.sort((event1, event2) => {
                    if (event1.closedAt != null && event2.closedAt == null) {
                        return 1;
                    }
                    if (event2.closedAt != null && event1.closedAt == null) {
                        return -1;
                    }
                    if (event1.createdAt < event2.createdAt) {
                        return 1;
                    }
                    if (event2.createdAt < event1.createdAt) {
                        return -1;
                    }
                    return 0;
                });
            },
            async getEvents() {
                const user = await userStore.getUser();

                eventStore.getUserEvents(user.id).onChange((events) => {
                    this.events = events;
                    this.sortByTime();
                });
            },
            getDateString(date) {
                return date !== null ? new Date(date).toLocaleString() : '';
            },
            goToEventCreation() {
                router.push('/events/create');
            },
            getFullName(person) {
                return person.firstName + ' ' + person.lastName;
            },
        },
        computed: {
            loaded() {
                return !!this.events;
            },
            filteredList() {
                return this.events.filter((event) => {
                    return event.title.toLowerCase().includes(this.search.toLowerCase()) ||
                        event.description.toLowerCase().includes(this.search.toLowerCase()) ||
                        event.owner.firstName.toLowerCase().includes(this.search.toLowerCase()) ||
                        event.owner.lastName.toLowerCase().includes(this.search.toLowerCase());
                });
            },
        },
    };
</script>

<style scoped>
    .limegreen {
        color: limegreen;
    }

    .form-control {
        width: 300px !important;
    }

    .card:hover {
        border: solid 1px limegreen;
    }

    .btn-limegreen {
        background-color: limegreen;
        border-color: limegreen;
    }

    .btn-limegreen:hover {
        background-color: #30b530;
        border-color: #30b530;
    }
</style>
