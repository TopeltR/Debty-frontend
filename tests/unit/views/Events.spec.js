import Events from '@/views/Events.vue'
import Vue from 'vue';
import {mount} from '@vue/test-utils';
import userStore from '@/stores/UserStore.ts';
import BootstrapVue from 'bootstrap-vue';

describe('Events', () => {
    Vue.use(BootstrapVue);
    console.error = jest.fn();
    console.warn = jest.fn();

    const get = jest.fn(() => Promise.resolve({data: [{id: 2}]}));
    const post = jest.fn();
    const del = jest.fn();
    Vue.prototype.$http = {get, post, delete: del};

    userStore.getUser = jest.fn(() => Promise.resolve({id: 1}));

    describe('getEvents', () => {
        it('should get all user events', async () => {
            const events = mount(Events);

            await events.vm.getEvents();

            expect(get).toHaveBeenCalledWith('/events/user/1');
            expect(events.vm.events).toEqual([{id: 2}]);
        })
    });

    it('should render', () => {
        const events = mount(Events);

        expect(events.html()).toMatchSnapshot();
    })
});
