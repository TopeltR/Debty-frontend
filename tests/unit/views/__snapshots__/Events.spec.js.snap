// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events should render 1`] = `
<div>
  <div class="container-fluid">
    <div class="row">
      <nav class="navbar fixed-top w-100 bg-gray py-0 h-md-60px navbar-light navbar-expand-sm">
        <div class="navbar-brand limegreen brand">debty</div> <button type="button" aria-label="Toggle navigation" aria-controls="nav_collapse" aria-expanded="false" class="navbar-toggler"><span class="navbar-toggler-icon"></span></button>
        <div id="nav_collapse" class="navbar-collapse collapse" style="display: none;">
          <ul class="navbar-nav h-60px mt-3 pb-2">
            <li class="nav-item hoverable-nav"><a target="_self" href="/home" class="nav-link">
                Home
              </a></li>
          </ul>
          <ul class="navbar-nav h-60px mt-3 pb-2">
            <li class="nav-item hoverable-nav"><a target="_self" href="/events" class="nav-link">
                Events
              </a></li>
          </ul>
          <ul class="navbar-nav h-60px mt-3 pb-2">
            <li class="nav-item hoverable-nav"><a target="_self" href="/debts" class="nav-link">
                Debts
                <!----></a></li>
          </ul>
          <ul class="navbar-nav h-60px mt-3 pb-2">
            <li class="nav-item hoverable-nav"><a target="_self" href="/contacts" class="nav-link">
                Contacts
                <!----></a></li>
          </ul>
          <ul class="navbar-nav h-60px mt-3 pb-2">
            <li class="nav-item hoverable-nav"><a target="_self" href="/profile" class="nav-link">
                Profile
              </a></li>
          </ul>
          <ul class="navbar-nav ml-auto h-60px mt-3 pb-2"><span class="d-none d-md-block mr-4 mt-2"> Welcome, <b></b></span>
            <li class="nav-item"><a target="_self" href="#" class="nav-link">
                <font-awesome-icon icon="sign-out-alt" class="icon mr-2"></font-awesome-icon> <span>Sign out</span>
              </a></li>
          </ul>
        </div>
      </nav>
    </div>
  </div>
  <div class="diagonal-background container-fluid">
    <div class="container">
      <div>
        <div><svg version="1.1" id="L7" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px" viewBox="0 0 100 100" enable-background="new 0 0 100 100" xml:space="preserve" class="height">
            <path fill="#fff" d="M31.6,3.5C5.9,13.6-6.6,42.7,3.5,68.4c10.1,25.7,39.2,38.3,64.9,28.1l-3.1-7.9c-21.3,8.4-45.4-2-53.8-23.3
c-8.4-21.3,2-45.4,23.3-53.8L31.6,3.5z">
              <animateTransform attributeName="transform" attributeType="XML" type="rotate" dur="2s" from="0 50 50" to="360 50 50" repeatCount="indefinite"></animateTransform>
            </path>
            <path fill="#fff" d="M42.3,39.6c5.7-4.3,13.9-3.1,18.1,2.7c4.3,5.7,3.1,13.9-2.7,18.1l4.1,5.5c8.8-6.5,10.6-19,4.1-27.7
c-6.5-8.8-19-10.6-27.7-4.1L42.3,39.6z">
              <animateTransform attributeName="transform" attributeType="XML" type="rotate" dur="1s" from="0 50 50" to="-360 50 50" repeatCount="indefinite"></animateTransform>
            </path>
            <path fill="#fff" d="M82,35.7C74.1,18,53.4,10.1,35.7,18S10.1,46.6,18,64.3l7.6-3.4c-6-13.5,0-29.3,13.5-35.3s29.3,0,35.3,13.5
L82,35.7z">
              <animateTransform attributeName="transform" attributeType="XML" type="rotate" dur="2s" from="0 50 50" to="360 50 50" repeatCount="indefinite"></animateTransform>
            </path>
          </svg></div>
      </div>
    </div>
  </div>
</div>
`;
